### PR TITLE
Fix outside of root element

### PR DESCRIPTION
**Pull request by alcortesm**
Merge #4 first.

Part of https://github.com/src-d/backlog/issues/610.

### CODE REVIEW

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 1]
```diff
@@ -2,33 +2,46 @@
...
+import java.util.Properties;
 
 public class ParserDemo {
+    private static final String debugKey = "bashsupport.debug"; // from BashParser
+
     public static void run(CharSequence code, BashVersion version) {
-        BashPsiBuilder b = builder(code, version);
-        FileParsing fp = new FileParsing();
+        boolean debug = "true".equals(System.getProperty(debugKey)) || "true".equals(System.getenv(debugKey));
+        System.out.println("debug = " + debug);
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Remove debug prints (by the way, you can use IntelliJ debugger for this).

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 2]
```diff
@@ -2,33 +2,46 @@
...
+
+        MockProject project = project();
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: This is an obsession of mine, but I recommend declaring every variable, method argument or class field as `final`, unless it requires to be reassigned for a good reason.

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 3]
```diff
@@ -2,33 +2,46 @@
...
+import com.intellij.lang.ParserDefinition;
 import com.intellij.lang.PsiBuilder;
 import com.intellij.lang.impl.PsiBuilderFactoryImpl;
 import com.intellij.mock.MockProject;
 import com.intellij.openapi.Disposable;
+import com.intellij.psi.tree.IElementType;
+
+import java.util.Properties;
 
 public class ParserDemo {
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Call this `BashParser`?

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 4]
```diff
@@ -2,33 +2,46 @@
...
+    private static final String debugKey = "bashsupport.debug"; // from BashParser
+
     public static void run(CharSequence code, BashVersion version) {
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: This should probably be `public ASTNode parse(CharSequence code, BashVersion version)`

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 5]
```diff
@@ -2,33 +2,46 @@
...
+        ParserDefinition parserDefinition = new BashParserDefinition();
 
-        fp.parseFile(b);
+        PsiBuilder builder = builder(parserDefinition, code);
+        IElementType root = parserDefinition.getFileNodeType();
 
-        System.out.println(b.getOriginalText());
+        ASTNode node = parser.parse(root, builder);
+
+        System.out.println(builder.getOriginalText());
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Remove debugging leftovers, use a JUnit test + debugger to test your stuff.

File: native/src/main/java/tech/sourced/ParserDemo.java [hunk 6]
```diff
@@ -2,33 +2,46 @@
...
 
-        ASTNode root = b.getTreeBuilt();
-        report(root);
+        report(node);
         System.out.flush();
     }
 
-    private static BashPsiBuilder builder(CharSequence code, BashVersion version) {
+    private static MockProject project() {
         com.intellij.openapi.extensions.ExtensionsArea rootArea = com.intellij.openapi.extensions.Extensions.getRootArea();
```
Comment (smola) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Use imports here for more readability.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 4 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Done.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 3 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Done.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 1 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Removed.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 2 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Thanks, I like the idea, I used to do the same with const in C++.

Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Done.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 5 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Done.

File: native/src/main/java/tech/sourced/ParserDemo.java (hunk 6 above)
Comment (alcortesm) [inline native/src/main/java/tech/sourced/ParserDemo.java]: Done.

**Pull request merged**